crypto/hash/keccak.go

```diff
@@ -1,14 +1,203 @@
-// license that can be found in the LICENSE file.
+package hash
 
-//go:build amd64 && !purego && gc
-// +build amd64,!purego,gc
+// Size returns the output size of the hash function in bytes.
+func (d *spongeState) Size() int {
+	return d.outputLen
+}
 
-package hash
+// Algorithm returns the hashing algorithm of the instance.
+func (s *spongeState) Algorithm() HashingAlgorithm {
+	return s.algo
+}
+
+// ComputeHash calculates and returns the digest of the input.
+// It updates the state (and therefore not thread-safe) and doesn't allow
+// further writing without calling Reset().
+func (s *spongeState) ComputeHash(data []byte) Hash {
+	s.Reset()
+	s.write(data)
+	return s.sum()
+}
+
+// SumHash returns the digest of the data written to the state.
+// It updates the state and doesn't allow further writing without
+// calling Reset().
+func (s *spongeState) SumHash() Hash {
+	return s.sum()
+}
+
+// Write absorbs more data into the hash's state.
+// It returns the number of bytes written and never errors.
+func (d *spongeState) Write(p []byte) (int, error) {
+	d.write(p)
+	return len(p), nil
+}
+
+// The functions below were copied and modified from golang.org/x/crypto/sha3.
+//
+// Copyright (c) 2009 The Go Authors. All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+const (
+	// maxRate is the maximum size of the internal buffer. SHA3-256
+	// currently needs the largest buffer among supported sponge-based
+	// algorithms.
+	maxRate = rateSHA3_256
+)
+
+type spongeState struct {
```

Comment:
my understanding is that `spongeState` is shared between different hashing algorithms (?) If that is the case, I would suggest putting it in its own file (e.g. `spongestate.go`). 

Maybe I am confused because `Keccak` is an entire family, to which sha3 also belongs?

Reply:
Yes, Keccak is the core base of the entire family. From that family, the crypto lib implements SHA3-256, SHA3-384, and Legacy Keccak-256. The sponge state is a common structure for all the Keccak-based family. That's why it is defined in the common file `keccak.go`.

Reply:
Calling the file `sponge.go` or `keccak.go` is basically equivalent. Sha3 uses Keccak (and therefore the sponge), and same for Legacy-Keccak, it uses Keccak (and the sponge).